packages/datetime/src/dateRangeInput.tsx

```diff
@@ -430,11 +487,93 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
         }
     }
 
+    private getDateRangeForCallback = (currValue?: moment.Moment, currBoundary?: DateRangeBoundary) => {
+        const otherBoundary = this.getOtherBoundary(currBoundary);
+        const otherValue = this.getStateKeysAndValuesForBoundary(otherBoundary).values.selectedValue;
+
+        const currDate = this.getDateForCallback(currValue);
+        const otherDate = this.getDateForCallback(otherValue);
+
+        return (currBoundary === DateRangeBoundary.START)
+            ? [currDate, otherDate]
+            : [otherDate, currDate];
+    }
+
+    private getDateForCallback = (momentDate: moment.Moment) => {
+        if (isMomentNull(momentDate)) {
+            return null;
+        } else if (!momentDate.isValid()) {
+            return new Date(undefined);
+        } else {
+            return fromMomentToDate(momentDate);
+        }
+    }
+
+    private getOtherBoundary = (boundary?: DateRangeBoundary) => {
+        return (boundary === DateRangeBoundary.START) ? DateRangeBoundary.END : DateRangeBoundary.START;
+    }
+
+    private doBoundariesOverlap(boundary: DateRangeBoundary, boundaryDate: moment.Moment) {
+        const otherBoundary = this.getOtherBoundary(boundary);
+        const otherBoundaryDate = this.getStateKeysAndValuesForBoundary(otherBoundary).values.selectedValue;
+
+        // TODO: add handling for allowSingleDayRange
+        if (boundary === DateRangeBoundary.START) {
+            return boundaryDate.isSameOrAfter(otherBoundaryDate);
+        } else {
+            return boundaryDate.isSameOrBefore(otherBoundaryDate);
+        }
+    }
+
     private isControlled = () => {
         return this.props.value !== undefined;
     }
 
+    private isInputEmpty = (inputString: string) => {
+        return inputString == null || inputString.length === 0;
+    }
+
+    private isInputInErrorState = (boundary: DateRangeBoundary) => {
+        const values = this.getStateKeysAndValuesForBoundary(boundary).values;
+        const { isInputFocused, inputString, selectedValue } = values;
+
+        const boundaryValue = (isInputFocused)
+            ? this.dateStringToMoment(inputString)
+            : selectedValue;
+
+        // break down the boolean logic to an elementary level to make it
+        // utterly simple to grok.
+
+        const isEmpty = isMomentNull(boundaryValue);
+        if (isEmpty) {
+            return false;
+        }
+
+        const isValid = boundaryValue.isValid();
+        if (!isValid) {
+            return true;
+        }
+
+        const isInRange = this.isMomentInRange(boundaryValue);
+        if (!isInRange) {
```

Comment:
actually then i'd even be fine with one big `if (A || B || C || D) { return true; }` situation. you can make it quite grokable with whitespace